File: core/internal/credential/credential.go
// SPDX-License-Identifier: MIT

// Package credential mints and parses shadownet-cred+jwt — the
// org_affiliation credential JWS the Issuer hands out in response to a
// valid CSR + ceremony approval.
//
// The credential's identifiers (iss, sub, org) are a union of three forms
// per RFC 0001 §6.1: a domain (Shadowname-mode issuer / org / subject's
// provider), a Shadowname (the attested subject), or a multibase Ed25519
// public key (direct-mode issuer, keyed Hub, or direct-mode subject).
//
// Spec: RFC 0001 §6.
//
// Schema mirror: shadownet-specs/schemas/credentials/credential.schema.json.
package credential

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/shadownet-protocol/shadownet/core/internal/crypto"
	"github.com/shadownet-protocol/shadownet/core/internal/identifiers"
	"github.com/shadownet-protocol/shadownet/core/internal/wellknown"
)

// Kind is the credential discriminator. v0.2 defines exactly one
// (RFC 0001 §6.2); future kinds slot in here.
type Kind string

const (
	// KindOrgAffiliation asserts the subject is a member of (or otherwise
	// acts for) the named organization.
	KindOrgAffiliation Kind = "org_affiliation"

	// MaxOrgAffiliationLifetime is the §6.3 cap on exp - iat for
	// org_affiliation credentials (30 days).
	MaxOrgAffiliationLifetime = 30 * 24 * time.Hour

	// DefaultLeeway is the ±60 s clock-skew tolerance per RFC 0001 §2.
	DefaultLeeway = 60 * time.Second
)

// Errors surfaced by mint / verify. Callers MAY wrap them with their own
// context (e.g. wellknown.ErrorURNPrefix + "creds_rejected" for HTTP) but
// the spec-level reason MUST stay attached.
var (
	ErrInvalid          = errors.New("credential: invalid")
	ErrUnknownKind      = errors.New("credential: unknown kind")
	ErrLifetimeExceeded = errors.New("credential: lifetime exceeds 30 days")
	ErrSignature        = errors.New("credential: signature did not verify")
	ErrExpired          = errors.New("credential: expired")
	ErrIATInFuture      = errors.New("credential: iat in the future")
	ErrIssuerUnauthd    = errors.New("credential: issuer not authorized for org")
)

// Revocation is the per-credential pointer into the issuer's status list
// (RFC 0001 §6.4). Both fields are required.
type Revocation struct {
	Epoch string `json:"epoch"`
	Idx   uint64 `json:"idx"`
}

// Payload is the decoded claims of a shadownet-cred+jwt.
type Payload struct {
	Iss  string     `json:"iss"`
	Sub  string     `json:"sub"`
	Kind Kind       `json:"kind"`
	Org  string     `json:"org"`
	Iat  int64      `json:"iat"`
	Exp  int64      `json:"exp"`
	Rev  Revocation `json:"rev"`
}

// Mint signs `p` with the issuer's key, producing a JWS-compact
// shadownet-cred+jwt. Per §6.3 only org_affiliation credentials are
// allowed and exp - iat MUST NOT exceed 30 days.
func Mint(p Payload, issuer crypto.KeyPair) (string, error) {
	if p.Kind != KindOrgAffiliation {
		return "", fmt.Errorf("%w: %q", ErrUnknownKind, p.Kind)
	}
	if p.Exp <= p.Iat {
		return "", fmt.Errorf("%w: exp must be greater than iat", ErrInvalid)
	}
	if time.Duration(p.Exp-p.Iat)*time.Second > MaxOrgAffiliationLifetime {
		return "", fmt.Errorf("%w: %ds", ErrLifetimeExceeded, p.Exp-p.Iat)
	}
	if err := validateIdentifier(p.Iss); err != nil {
		return "", fmt.Errorf("%w: iss: %v", ErrInvalid, err)
	}
	if err := validateIdentifier(p.Sub); err != nil {
		return "", fmt.Errorf("%w: sub: %v", ErrInvalid, err)
	}
	if err := validateIdentifier(p.Org); err != nil {
		return "", fmt.Errorf("%w: org: %v", ErrInvalid, err)
	}
	if p.Rev.Epoch == "" {
		return "", fmt.Errorf("%w: rev.epoch required", ErrInvalid)
	}
	pubMB, err := identifiers.EncodePubKey(issuer.Public)
	if err != nil {
		return "", fmt.Errorf("credential: encode issuer pubkey: %w", err)
	}
	return crypto.SignJWT(issuer.Private, p, crypto.SignerOptions{
		Type:  wellknown.TypShadownetCredJWT,
		KeyID: kidFromIssuer(p.Iss, pubMB),
	})
}

// kidFromIssuer chooses a kid header value based on the issuer
// identifier's form. Domain-mode issuers use the AgentCard-style
// "shadownet@<domain>" kid; keyed issuers use the multibase pubkey form
// directly (matching the §5.3 direct-mode kid convention).
func kidFromIssuer(iss, issuerPubMultibase string) string {
	switch identifiers.Classify(iss) {
	case identifiers.ClassPubKey:
		return iss
	case identifiers.ClassDomain:
		return "shadownet@" + iss
	default:
		// Shadowname-issued credentials are not defined by §6, but fall
		// back to the pubkey form for safety.
		return issuerPubMultibase
	}
}

// VerifyOptions threads the resolver callbacks the caller controls (DNS
// lookup of the issuer's domain provider record, status-list fetch, etc.).
type VerifyOptions struct {
	// Now overrides time.Now for deterministic testing.
	Now func() time.Time

	// Leeway tolerates iat/exp clock skew. Defaults to DefaultLeeway when
	// zero.
	Leeway time.Duration

	// ResolveIssuerKey returns the verification key for the given iss
	// identifier. Caller-supplied so the credential package stays free of
	// DNS / HTTP plumbing. For keyed issuers, the identifier IS the key;
	// callers MAY just return identifiers.DecodePubKey(iss) directly.
	ResolveIssuerKey func(iss string) (ed25519.PublicKey, error)

	// AuthorizeIssuerForOrg implements §6.6. The default rules: iss == org;
	// iss is a sub-domain of org (only meaningful when both are domains);
	// iss appears under delegate= in `_shadownet.<org>`'s TXT record (only
	// meaningful when org is a domain). The keyed-issuer carve-out
	// (iss == org only) is enforced by the implementation supplied here.
	AuthorizeIssuerForOrg func(iss, org string) error
}

// Verify parses, validates, and authenticates a shadownet-cred+jwt against
// the verification key returned by opts.ResolveIssuerKey. It implements
// §6 validation steps 1-6 (header typ + alg, payload shape, lifetime,
// signature, issuer-for-org authorization). Revocation (step 7) and
// trust-store evaluation (step 8) are left to the caller; see RFC 0001 §6,
// §7.
func Verify(token string, opts VerifyOptions) (Payload, error) {
	hdr, err := crypto.PeekHeader(token)
	if err != nil {
		return Payload{}, fmt.Errorf("%w: header: %v", ErrInvalid, err)
	}
	if hdr.Typ != wellknown.TypShadownetCredJWT {
		return Payload{}, fmt.Errorf("%w: typ %q, want %q", ErrInvalid, hdr.Typ, wellknown.TypShadownetCredJWT)
	}
	if hdr.Alg != "EdDSA" {
		return Payload{}, fmt.Errorf("%w: alg %q, want EdDSA", ErrInvalid, hdr.Alg)
	}

	// Decode claims unverified first so we can resolve the issuer key.
	var raw struct {
		Iss  string          `json:"iss"`
		Sub  string          `json:"sub"`
		Kind string          `json:"kind"`
		Org  string          `json:"org"`
		Iat  int64           `json:"iat"`
		Exp  int64           `json:"exp"`
		Rev  json.RawMessage `json:"rev"`
	}
	parts, payload, err := peekClaims(token)
	if err != nil {
		return Payload{}, err
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return Payload{}, fmt.Errorf("%w: payload: %v", ErrInvalid, err)
	}
	if raw.Kind != string(KindOrgAffiliation) {
		return Payload{}, fmt.Errorf("%w: %q", ErrUnknownKind, raw.Kind)
	}
	var rev Revocation
	if len(raw.Rev) == 0 {
		return Payload{}, fmt.Errorf("%w: rev required", ErrInvalid)
	}
	if err := json.Unmarshal(raw.Rev, &rev); err != nil {
		return Payload{}, fmt.Errorf("%w: rev: %v", ErrInvalid, err)
	}
	if rev.Epoch == "" {
		return Payload{}, fmt.Errorf("%w: rev.epoch required", ErrInvalid)
	}
	if raw.Exp <= raw.Iat {
		return Payload{}, fmt.Errorf("%w: exp must be greater than iat", ErrInvalid)
	}
	if time.Duration(raw.Exp-raw.Iat)*time.Second > MaxOrgAffiliationLifetime {
		return Payload{}, ErrLifetimeExceeded
	}
	if err := validateIdentifier(raw.Iss); err != nil {
		return Payload{}, fmt.Errorf("%w: iss: %v", ErrInvalid, err)
	}
	if err := validateIdentifier(raw.Sub); err != nil {
		return Payload{}, fmt.Errorf("%w: sub: %v", ErrInvalid, err)
	}
	if err := validateIdentifier(raw.Org); err != nil {
		return Payload{}, fmt.Errorf("%w: org: %v", ErrInvalid, err)
	}

	now := time.Now
	if opts.Now != nil {
		now = opts.Now
	}
	leeway := opts.Leeway
	if leeway == 0 {
		leeway = DefaultLeeway
	}
	nowT := now()
	if nowT.Unix() > raw.Exp+int64(leeway/time.Second) {
		return Payload{}, ErrExpired
	}
	if nowT.Unix()+int64(leeway/time.Second) < raw.Iat {
		return Payload{}, ErrIATInFuture
	}

	if opts.ResolveIssuerKey == nil {
		return Payload{}, fmt.Errorf("%w: VerifyOptions.ResolveIssuerKey is nil", ErrInvalid)
	}
	pub, err := opts.ResolveIssuerKey(raw.Iss)
	if err != nil {
		return Payload{}, fmt.Errorf("credential: resolve issuer key for %q: %w", raw.Iss, err)
	}
	if _, _, err := crypto.VerifyJWS(pub, parts); err != nil {
		return Payload{}, fmt.Errorf("%w: %v", ErrSignature, err)
	}

	if opts.AuthorizeIssuerForOrg != nil {
		if err := opts.AuthorizeIssuerForOrg(raw.Iss, raw.Org); err != nil {
			return Payload{}, fmt.Errorf("%w: %v", ErrIssuerUnauthd, err)
		}
	}

	return Payload{
		Iss:  raw.Iss,
		Sub:  raw.Sub,
		Kind: Kind(raw.Kind),
		Org:  raw.Org,
		Iat:  raw.Iat,
		Exp:  raw.Exp,
		Rev:  rev,
	}, nil
}

// peekClaims returns the token (unchanged) and its base64url-decoded
// payload segment without verifying the signature.
func peekClaims(token string) (string, []byte, error) {
	segs := strings.Split(token, ".")
	if len(segs) != 3 {
		return "", nil, fmt.Errorf("%w: malformed compact serialization", ErrInvalid)
	}
	payload, err := base64.RawURLEncoding.DecodeString(segs[1])
	if err != nil {
		return "", nil, fmt.Errorf("%w: payload base64: %v", ErrInvalid, err)
	}
	return token, payload, nil
}

// validateIdentifier accepts any of the three union forms.
func validateIdentifier(s string) error {
	if identifiers.Classify(s) == identifiers.ClassUnknown {
		return fmt.Errorf("%w: %q is not a domain, shadowname, or multibase pubkey", ErrInvalid, s)
	}
	return nil
}

File: python-sdk/src/shadownet/credential.py
"""Credentials — RFC 0001 §6.

One credential kind in v0.2: ``org_affiliation``. Plain JWS-compact JWT, no
W3C VC wrapping. Schema mirrors
``shadownet-specs/schemas/credentials/credential.schema.json``.

Verification is split into three concerns so callers can inject them:

  * **Issuer-key resolution** (``resolve_issuer_key``): given an issuer
    domain, return the multibase Ed25519 key. The default looks the key up
    via the provider DNS record (§4.2).
  * **Issuer-org authorization** (``check_issuer_authorized_for_org``):
    apply the §6.6 rules. Same-domain and sub-domain checks are local;
    DNS-delegate checks require a TXT lookup, which the default does.
  * **Revocation** (``check_revoked``): consult the status list (§6.4).
    Delegated to ``shadownet.status``.
"""

from __future__ import annotations

import base64
import time
from typing import TYPE_CHECKING, Annotated, Final

from pydantic import BaseModel, ConfigDict, Field, ValidationError

if TYPE_CHECKING:
    from collections.abc import Callable

from shadownet.crypto.ed25519 import Ed25519KeyPair, SignatureError
from shadownet.crypto.jwt import (
    JWTError,
    decode_header,
    decode_unverified_claims,
    sign_jwt,
)
from shadownet.errors import ShadownetError
from shadownet.identifiers import (
    Identifier,
    IssuerIdentifier,
    is_public_key_identifier,
    is_subdomain_of,
    parse_public_key,
)
from shadownet.provider import (
    ProviderResolutionError,
    lookup_provider_record,
)

__all__ = [
    "CREDENTIAL_TYP",
    "MAX_LIFETIME_SECONDS",
    "ORG_AFFILIATION",
    "CredentialError",
    "CredentialPayload",
    "RevocationPointer",
    "VerifiedCredential",
    "default_issuer_authorization_check",
    "default_issuer_key_resolver",
    "mint_credential",
    "verify_credential",
]


CREDENTIAL_TYP: Final = "shadownet-cred+jwt"
ORG_AFFILIATION: Final = "org_affiliation"
# §6.3: org_affiliation max lifetime is 30 days.
MAX_LIFETIME_SECONDS: Final = 30 * 24 * 60 * 60
# §2: all time comparisons tolerate ±60 s of skew.
DEFAULT_LEEWAY_SECONDS: Final = 60


class CredentialError(ShadownetError):
    """A credential failed to parse, verify, or satisfy policy."""


class RevocationPointer(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)
    epoch: str = Field(min_length=1)
    idx: int = Field(ge=0)


class CredentialPayload(BaseModel):
    """Decoded payload of a ``shadownet-cred+jwt``.

    ``iss`` and ``org`` accept either a domain (Shadowname-mode issuer / org)
    or a multibase Ed25519 public key (keyed issuer / Hub). ``sub`` accepts
    a Shadowname or a public key (direct-mode Shadow).
    """

    model_config = ConfigDict(extra="allow", frozen=True, populate_by_name=True)

    iss: IssuerIdentifier
    sub: Identifier
    kind: Annotated[str, Field(pattern="^[a-z_]+$")]
    org: IssuerIdentifier
    iat: int = Field(ge=0)
    exp: int = Field(ge=0)
    rev: RevocationPointer


class VerifiedCredential(BaseModel):
    """A credential whose signature, claims, lifetime, and §6.6 authorization
    have been checked. Revocation and trust-store policy are layered on by
    the caller (see ``shadownet.trust``)."""

    model_config = ConfigDict(frozen=True)
    payload: CredentialPayload
    issuer_key: str
    raw_jws: str


def mint_credential(
    payload: CredentialPayload,
    issuer_key: Ed25519KeyPair,
) -> str:
    if payload.exp - payload.iat > MAX_LIFETIME_SECONDS:
        raise CredentialError(
            f"{ORG_AFFILIATION} lifetime exceeds 30 days (exp - iat = {payload.exp - payload.iat}s)"
        )
    if payload.kind != ORG_AFFILIATION:
        raise CredentialError(f"unknown credential kind {payload.kind!r}")
    claims = payload.model_dump()
    return sign_jwt(claims, issuer_key, header_extras={"typ": CREDENTIAL_TYP})


def verify_credential(
    token: str,
    *,
    now: int | None = None,
    leeway: int = DEFAULT_LEEWAY_SECONDS,
    resolve_issuer_key: Callable[[str], str] | None = None,
    check_issuer_authorized_for_org: Callable[[str, str], None] | None = None,
) -> VerifiedCredential:
    """Run §6 validation steps 1-6 against ``token``.

    Steps 7 (revocation) and 8 (trust store) are not the responsibility of
    this function; see ``shadownet.status`` and ``shadownet.trust``.
    """
    try:
        header = decode_header(token)
    except JWTError as exc:
        raise CredentialError(f"invalid JWS: {exc}") from exc
    if header.get("typ") != CREDENTIAL_TYP:
        raise CredentialError(f"typ must be {CREDENTIAL_TYP!r}, got {header.get('typ')!r}")
    if header.get("alg") != "EdDSA":
        raise CredentialError(f"alg must be EdDSA, got {header.get('alg')!r}")

    try:
        unverified = decode_unverified_claims(token)
    except JWTError as exc:
        raise CredentialError(f"unable to decode claims: {exc}") from exc

    try:
        payload = CredentialPayload.model_validate(unverified)
    except ValidationError as exc:
        raise CredentialError(f"credential payload invalid: {exc}") from exc

    if payload.kind != ORG_AFFILIATION:
        raise CredentialError(f"unknown credential kind {payload.kind!r}")

    if payload.exp - payload.iat > MAX_LIFETIME_SECONDS:
        raise CredentialError("credential lifetime exceeds 30 days")

    current = int(time.time()) if now is None else now
    if payload.exp < current - leeway:
        raise CredentialError("credential expired")
    if payload.iat > current + leeway:
        raise CredentialError("credential iat in the future")

    resolver = resolve_issuer_key or default_issuer_key_resolver
    issuer_key_multibase = resolver(payload.iss)
    issuer_pk_bytes = parse_public_key(issuer_key_multibase)
    issuer_key = Ed25519KeyPair.from_public_bytes(issuer_pk_bytes)

    try:
        _verify_jws_signature(token, issuer_key)
    except (JWTError, SignatureError) as exc:
        raise CredentialError(f"signature verification failed: {exc}") from exc

    authorize = check_issuer_authorized_for_org or default_issuer_authorization_check
    authorize(payload.iss, payload.org)

    return VerifiedCredential(payload=payload, issuer_key=issuer_key_multibase, raw_jws=token)


def default_issuer_key_resolver(issuer: str) -> str:
    """Resolve the issuer's signing key.

    Keyed issuers (RFC 0001 §3.3): ``iss`` IS the verification key, returned
    verbatim. Domain issuers: DNS-resolve ``_shadownet.<iss>`` TXT (§4.2) and
    return the published ``pk``.
    """
    if is_public_key_identifier(issuer):
        return issuer
    try:
        record = lookup_provider_record(issuer)
    except ProviderResolutionError as exc:
        raise CredentialError(f"could not resolve issuer {issuer!r}: {exc}") from exc
    if not record.provider_keys:
        raise CredentialError(f"issuer {issuer!r} has no provider key")
    return record.provider_keys[0]


def default_issuer_authorization_check(issuer: str, org: str) -> None:
    """Apply RFC 0001 §6.6.

    Rule 1 — ``iss == org`` — is the **only** path open to keyed issuers and
    the trivial case for domain issuers. Rules 2 (sub-domain) and 3 (DNS
    delegate) are domain-only.
    """
    if issuer == org:
        return
    if is_public_key_identifier(issuer) or is_public_key_identifier(org):
        raise CredentialError(
            f"keyed issuer {issuer!r} not authorized to attest for org {org!r} "
            "(only iss == org accepted for keyed issuers per §6.6)"
        )
    if is_subdomain_of(issuer, org):
        return
    try:
        org_record = lookup_provider_record(org)
    except ProviderResolutionError as exc:
        raise CredentialError(f"could not verify issuer authorization for {org!r}: {exc}") from exc
    if issuer.lower() in (d.lower() for d in org_record.delegates):
        return
    raise CredentialError(f"issuer {issuer!r} is not authorized to attest for org {org!r}")


def _verify_jws_signature(token: str, issuer_key: Ed25519KeyPair) -> None:
    parts = token.split(".")
    if len(parts) != 3:
        raise JWTError("malformed JWS compact serialization")
    signing_input = (parts[0] + "." + parts[1]).encode("ascii")
    sig = base64.urlsafe_b64decode(parts[2] + "=" * (-len(parts[2]) % 4))
    issuer_key.verify(sig, signing_input)
